File: sdk/src/server-api/sc/networking/clients/GameLoaderClient.java
package sc.networking.clients;

import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sc.api.plugins.IGameState;
import sc.networking.FileSystemInterface;
import sc.protocol.ProtocolPacket;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * This client serves the purpose to load a game state from
 * any XML file (a replay for example).
 * It is used to load a given board to play on.
 */
public final class GameLoaderClient extends XStreamClient implements IPollsHistory {
  private static Logger logger = LoggerFactory.getLogger(GameLoaderClient.class);

  private List<IHistoryListener> listeners = new ArrayList<>();

  public GameLoaderClient(InputStream inputStream) throws IOException {
    super(new FileSystemInterface(inputStream));
    logger.trace("Loading game from {}", inputStream);
  }

  @Override
  protected void onObject(@NotNull ProtocolPacket message) {
    for (IHistoryListener listener : this.listeners) {
      listener.onNewState(null, (IGameState) message);
    }
  }

  @Override
  public void addListener(IHistoryListener listener) {
    this.listeners.add(listener);
  }

  @Override
  public void removeListener(IHistoryListener listener) {
    this.listeners.remove(listener);
  }

}

File: sdk/src/server-api/sc/networking/clients/GameLoaderClient.kt
package sc.networking.clients

import org.slf4j.LoggerFactory
import sc.api.plugins.IGameState
import sc.networking.FileSystemInterface
import sc.protocol.ProtocolPacket
import sc.protocol.room.MementoMessage
import sc.protocol.room.RoomPacket
import sc.shared.GameResult
import java.io.File
import java.io.IOException
import java.io.InputStream

/**
 * This client serves the purpose to load a game state from
 * any XML file (a replay for example).
 * It is used to load a given board to play on.
 */
class GameLoaderClient(inputStream: InputStream): XStreamClient(FileSystemInterface(inputStream)) {
    constructor(file: File): this(file.inputStream())
    
    private val history: MutableList<IGameState> = ArrayList(50)
    var result: GameResult? = null
        private set
    
    override fun onObject(message: ProtocolPacket) {
        logger.trace("Adding packet to replay: {}", message)
        if (message !is RoomPacket)
            throw IOException("Can't extract replay from $message")
        when (val msg = message.data) {
            is MementoMessage -> history.add(msg.state)
            is GameResult -> result = msg
            else -> logger.debug("Unknown message in replay: {}", msg)
        }
    }
    
    fun getHistory(): List<IGameState> {
        start()
        while (!isClosed)
            Thread.sleep(100)
        return history
    }
    
    fun getTurn(turn: Int) =
        getHistory().first {
            it.turn >= turn
        }
    
    companion object {
        private val logger = LoggerFactory.getLogger(this::class.java)
    }
}